src/server/cluster/outgoing_slot_migration.cc

```diff
@@ -17,46 +18,58 @@ class OutgoingMigration::SliceSlotMigration {
   SliceSlotMigration(DbSlice* slice, SlotSet slots, uint32_t sync_id, journal::Journal* journal,
                      Context* cntx, io::Sink* dest)
       : streamer_(slice, std::move(slots), sync_id, journal, cntx) {
-    streamer_.Start(dest);
     state_.store(MigrationState::C_FULL_SYNC, memory_order_relaxed);
+    sync_fb_ = Fiber("slot-snapshot", [this, dest] { streamer_.Start(dest); });
   }
 
   void Cancel() {
     streamer_.Cancel();
   }
 
+  void WaitForSnapshotFinished() {
+    sync_fb_.JoinIfNeeded();
+  }
+
   void Finalize() {
     streamer_.SendFinalize();
     state_.store(MigrationState::C_FINISHED, memory_order_relaxed);
   }
 
   MigrationState GetState() const {
-    auto state = state_.load(memory_order_relaxed);
-    return state == MigrationState::C_FULL_SYNC && streamer_.IsSnapshotFinished()
-               ? MigrationState::C_STABLE_SYNC
-               : state;
+    return state_.load(memory_order_relaxed);
   }
 
  private:
   RestoreStreamer streamer_;
   // Atomic only for simple read operation, writes - from the same thread, reads - from any thread
   atomic<MigrationState> state_ = MigrationState::C_CONNECTING;
+  Fiber sync_fb_;
 };
 
 OutgoingMigration::OutgoingMigration(std::uint32_t flows_num, std::string ip, uint16_t port,
                                      SlotRanges slots, Context::ErrHandler err_handler)
     : host_ip_(ip), port_(port), slots_(slots), cntx_(err_handler), slot_migrations_(flows_num) {
 }
 
-OutgoingMigration::~OutgoingMigration() = default;
+OutgoingMigration::~OutgoingMigration() {
+  main_sync_fb_.JoinIfNeeded();
+}
 
 void OutgoingMigration::StartFlow(DbSlice* slice, uint32_t sync_id, journal::Journal* journal,
                                   io::Sink* dest) {
   const auto shard_id = slice->shard_id();
 
-  std::lock_guard lck(flows_mu_);
-  slot_migrations_[shard_id] =
-      std::make_unique<SliceSlotMigration>(slice, slots_, sync_id, journal, &cntx_, dest);
+  MigrationState state = MigrationState::C_NO_STATE;
+  {
+    std::lock_guard lck(flows_mu_);
+    slot_migrations_[shard_id] =
+        std::make_unique<SliceSlotMigration>(slice, slots_, sync_id, journal, &cntx_, dest);
+    state = GetStateImpl();
+  }
+
+  if (state == MigrationState::C_FULL_SYNC) {
+    main_sync_fb_ = Fiber("outgoing_migration", &OutgoingMigration::SyncFb, this);
```

Comment:
why do you need the main sync fb?

Reply:
To know when snapshot sending is finished. After that I can initiate Finalize cmd automatically